### DISCUSSION TITLE
How to get a human readable trace of forward pass of model w/ ops, shapes, etc

### DISCUSSION THREAD

**Discussion by rwightman** [google/flax#546] [Q&A]
Once you have a model defined in Flax, is there any way to get a human readable text (or visualizable graph) of the model structure, sequence of operations in the forward pass? Ideally something that is relatable to the model definition with some layer name associations.

So far working with Flax I've found it akward not being able to inspect the model structure easily. I had a little bug where I failed to pass the correct act fn to a module in a larger model that would have been really easy to spot with a readable dump of the forward op chain.

It seems like this should be possible, if not already present in jax or flax. I scanned through Flax / Jax code/issue tracker/docs and didn't see any mention of such a functionality that stood out. I don't really want the output of the XLA compiler or a profiler. Ideally something higher level that can be used to quickly debug a model (ie do a forward pass with a debug flag enabled, set a context manager, etc). Working in TF I used to poke through the graph in tensorboard, in PyTorch when written in modular style it's easy to inspect/print/debug inspect the module hierarchy.

**Comment by jheek** [comment]
This is possible with the (currently experimental) `flax.linen` api. As in the representation of a `flax.linen.Module` should show the nested modules and their hyper paramaters similar to PyTorch.

The `flax.nn` api does not store the hyper-paramaters on the object so this makes it harder to inspect in Python indeed. This is own reason for why we are moving away from this approach.

If you want something that is more at the Tensorboard graph level you can also use `jax.make_jaxpr(some_fn)(*args)` which will print out all the JAX operations executed by that function.

**Comment by rwightman** [comment]
@jheek I'm using Linen but went with the @compact `__call__` style over using setup() as it seemed to be featured more in the examples, that didn't result in any inspectable structure. 

If I were to use setup and be able to inspect the module hierarchy that could help, but like Objax there is still an area where a higher level trace/xpr output that can inject module boundaries and layer/weight names would be very useful. In most JAX abstractions, all of the functional (ie activations) are left as functions. There is no parallel set of modules as you have in PyTorch. So walking modules would still exclude activations.

The make_jaxpr is something, it it is quite low level and loses any layer bounds and naming that would relate to the original model.

**Comment by jheek** [comment]
I think it would be good to have an inspection method that also takes into account modules that were created inside a @compact method. Do you want to make a PR for that?

**Reply by rwightman** [reply]
@jheek I'll dig in a bit more and see if I can come up with a clean way of doing that. I'll bring up any proposals here before creating a PR.

**Reply by jheek** [reply]
Sorry I meant to say do you want to create an issue. Of course you are welcoming to fill a PR but I think this is quite important and the core team will look into this if you just file an issue.

**Reply by mattiasmar** [reply]
@jheek @rwightman Was this thread ever concluded in a PR? I too am looking for a way to extract the computational graph from JAX or FLAX. Any pointers?
